### ISSUE TITLE
async.py stop thread doesn't die on exit

### ISSUE THREAD

**Issue by orcaomar** [GochoMugo/firecall#5]
in the async file, there is this code for "stop":

    def stop(self, timeout=0):
        self.stopwatch = Thread(target=self.__stopper, args=(timeout,)).start()

unfortunately, if you make your timeout VERY long, then even if you make call stop with a timeout of 0, that will stop the request thread but the first watch, with the very long timeout, will have an outstanding thread.

then, it's very hard to close the program. python won't exit with that thread hanging about.

here is one solution. 

    def stop(self, timeout=0):
        t = Thread(target=self.__stopper, args=(timeout,))
        t.daemon = True
        self.stopwatch = t.start() #Thread(target=self.__stopper, args=(timeout,)).start()

namely, you make the thread that's waiting to STOP the request thread actually watch for exit. this works, i tried it.

anyway, i'm open to other ideas.

**Comment by GochoMugo** [comment]
I have had some time and researched on **threading** in Python. I have been through the Python docs and it seems we could implement the watch using [Timer Objects](https://docs.python.org/2/library/threading.html#timer-objects). A timer object has a `.cancel` method that can be used to cancel a timer object as it is in waiting state.

What do you think?

**Comment by orcaomar** [comment]
i thought about your timer idea, and it seems like more work than is needed. i think we should go with my solution -- essentially the only change is to make the stopwatch thread a daemon thread. i think this is the right solution after reading the answer to this question:

http://stackoverflow.com/questions/190010/daemon-threads-explanation

because the stopwatch threads should basically be forgotten about when the program is shutting down. i don't think there's any need to do any cleanup. do you?

**Comment by orcaomar** [comment]
however, the request thread shouldn't be daemonized since who knows what user callbacks might be running on the request thread. for that thread to be shutdown correctly, i've just been calling stop(0) to make sure it exits properly. perhaps you want to add a close() call which will make sure the request thread is properly ended? then, when exiting the program, i would just call something like:

on_change =  my_firebase.onChange(...)

...
# when exiting:
on_change.close()

whose implementation might be something like:


class async:
...
  def close(self):
    self.stop(0)
...

**Comment by GochoMugo** [comment]
The problem lies here:

```python
    # Stop Watch: This will kill the Watch associated with this Instance
    def stop(self, timeout=0):
        self.stopwatch = Thread(target=self.__stopper, args=(timeout,)).start()

    # The Stopper
    def __stopper(self, timeout):
        self.__event.wait(timeout)
        self.__watch = False
```

Should `timeout` be long the `self.stopwatch` thread will be stuck waiting on `self.__event` to be `set()` or until the `timeout` occurs. A subsequent attempt at stopping the watch by invoking `stop(0)` certainly kills the request thread **but**  keeps the initial `self.stopwatch` thread running.

It is **not** advised to kill a thread, although in our case, killing the `self.stopwatch` thread would do little harm. See this [discussion](http://stackoverflow.com/questions/323972/is-there-any-way-to-kill-a-thread-in-python). A better and simpler way would be:

```python
    # The Stopper
    def __stopper(self, timeout):
        self.__event.wait(timeout)
        self.__watch = False
        self.__event.set() # this stops any previously-created thread objects for stopping the watch
```

What do you think?

**Comment by orcaomar** [comment]
that seems fine, but any future thread that waits will not wait because the flag is set. so, you would change this code:

```python
    # Stop Watch: This will kill the Watch associated with this Instance
    def stop(self, timeout=0):
        self.stopwatch = Thread(target=self.__stopper, args=(timeout,)).start()
```

to 
```python
    # Stop Watch: This will kill the Watch associated with this Instance
    def stop(self, timeout=0):
        self.__event.clear()
        self.stopwatch = Thread(target=self.__stopper, args=(timeout,)).start()
```

and that should work. if you are not going to add a "close" method, i would recommend documenting that calling `stop()` is needed to properly terminate the running threads.

**Comment by GochoMugo** [comment]
Invoking `stop()` again once the watch has stopped should **not** wait since the request thread will have already stopped. There will be **no** point of waiting anymore.

The point of **documenting** is a great one. We should tell users that calling `stop()` is required to ensure no threads remaining hanging when their program exits.

**Comment by orcaomar** [comment]
ah i see what you're saying -- you're saying that once the timer is stopped, this instance of the onchange call is no longer usable -- because the request thread has exited. so there's no point clearing -- once it has actually STOPPED, you can't use it again. 

great, look forward to the change! thank you

**Comment by GochoMugo** [comment]
Would you wish to make the change and send a pull request? :smile:

**Comment by GochoMugo** [comment]
Thank you @orcaomar!!! Fixed in v1.0.1!!! :+1:
